Task3

LesterOper opened the pull request.

LesterOper committed:
Smth wrong with session

LesterOper committed:
Add files via upload

LesterOper committed:
Add files via upload

LesterOper committed:
Add files via upload

LesterOper committed:
Finished Task
Implemented HttpSessionCart
Implemented Cart, CartService, CartItem
Finished task "Self check"

SergeiStarikov commented on src/main/java/com/es/phoneshop/Cart/Cart.java:
```diff
@@ -1,9 +1,16 @@
+        cart = new ArrayList<>();
+    }
+    
+    public List<CartItem> getList(){
```
Here and in other places - coding style guides:

getList(){ --> getList() {
if(session.getAttribute(VIEWED_ITEMS)==null){ --> if (session.getAttribute(VIEWED_ITEMS) == null) {

Style guides: https://google.github.io/styleguide/javaguide.html

Let IDEA format the file for you: https://www.jetbrains.com/help/idea/reformat-file-dialog.html

Ctrl + Shift + Alt + L or appropriate option in settings

SergeiStarikov commented on src/main/java/com/es/phoneshop/Cart/HttpSessionCartServiceService.java:
```diff
@@ -0,0 +1,67 @@
+    public void addProduct(Cart cartList,Product product, int quantity) {
+        if(cartList.getList().stream().anyMatch(prod-> prod.getProduct().getId().equals(product.getId()))){
+            cartList.getList().stream().
+                    filter(prod-> prod.getProduct().getId().equals(product.getId())).
```
Streams formatting:

 cartList.getList().stream()
                    .filter()
                    .map()

SergeiStarikov commented on src/main/webapp/WEB-INF/pages/productDescription.jsp:
```diff
@@ -15,7 +13,7 @@
   <p>
     <c:if test="${error!=null}">
-        <font color ="red"> <c:out value="You enter smth wrong"/></font>
+        <span style="color: red; "><c:out value="You enter smth wrong"/></span>
```
Here and in other places - HTML formatting:
<span>
    <...>
</span>

SergeiStarikov commented on src/main/webapp/WEB-INF/pages/productDescription.jsp:
```diff
@@ -38,24 +36,45 @@
-            </font>
+            <c:choose>
+                <c:when test="${error!=null}">
+                    <span style="color: red; "><c:out value="${error}"/></span>
```
Please move inline styles to .css file

SergeiStarikov commented on src/main/webapp/WEB-INF/pages/productDescription.jsp:
```diff
@@ -38,24 +36,45 @@
+            </table>
+            </p>
+            </c:forEach>
+        </c:if>
```
Formatting again:

Compare what it takes to read this:

```
<c:set var="list" value="${sessionScope['cartList'].list}"/>
        <c:if test="${list!=null}">
            <c:forEach items="${list}" var="cart" >
                <p><c:out value="${cart}"/></p>
            </c:forEach>
        </c:if>
    <c:set var="list" value="${sessionScope['viewed'].list}"/>
        <c:if test="${list!=null}">
            <c:forEach items="${list}" var="view">
            <p>
                ...
            </p>
            </c:forEach>
        </c:if>
```

and this:

```
<c:set var="list" value="${sessionScope['cartList'].list}"/>
<c:set var="list" value="${sessionScope['viewed'].list}"/>

<c:if test="${list!=null}">
    <c:forEach items="${list}" var="cart">
         <p>
             <c:out value="${cart}"/>
         </p>
    </c:forEach>
</c:if>

<c:if test="${list!=null}">
    <c:forEach items="${list}" var="view">
        <p>
            ...        
        </p>
    </c:forEach>
<c:if>
```

SergeiStarikov commented on src/main/webapp/WEB-INF/pages/productDescription.jsp:
```diff
@@ -38,24 +36,45 @@
+    <c:set var="list" value="${sessionScope['viewed'].list}"/>
+        <c:if test="${list!=null}">
+            <c:forEach items="${list}" var="view">
+            <p>
```
```
<p>
            <table>
                <td>
                   <p>
```

Do we need this? table inside p, p inside td. No <tr> tag

Please check proper table layout: https://www.w3schools.com/html/html_tables.asp

Please watch for IntelliJ IDEA hints (yellow marks on the right, colored strings) - they can help find errors.

LesterOper committed:
Fixed sort by price and field
Formatted by comments
Recently viewed products are on productDescription.jsp
Added some tests (HttpSessionCartServiceTest::testAddToCart and testDoPost may be right)

LesterOper committed:
Here test for HttpSessionCartService
Forgot to add this test in last commit

SergeiStarikov commented on src/test/java/com/es/phoneshop/Session/HttpSessionCartServiceTest.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testAddProductToCart() {
```
Please add more tests for edge cases: add product with quantity 1000, etc.

SergeiStarikov commented on src/main/webapp/WEB-INF/pages/productDescription.jsp:
```diff
@@ -5,76 +5,85 @@
+    <c:if test="${list!=null}">
+        <c:forEach items="${list}" var="view">
             <table>
                 <td>
```
Formatting: no <tr> tag

SergeiStarikov commented on src/main/webapp/WEB-INF/pages/productDescription.jsp:
```diff
@@ -5,76 +5,85 @@
+        </tr>
+    </table>
+    <form method="post">
+        <label>Quantity</label><input type="text" name="quantity"
```
HTML formatting
